Add VolumeData Python Wrapper Class

isaaclegred opened the pull request:
## Proposed changes

This PR extends the spectre python hdf5 interface to the `VolumeData` class, i.e. it adds a `VolumeData` python class and additional functions to the H5File python class which are needed to insert and access `.vol` files inside of h5 files.    

### Types of changes:

- [ ] Bugfix
- [x ] New feature

### Component:

- [x ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

isaaclegred committed:
Add Python wrapper for H5 VolumeData class

nilsdeppe commented on src/IO/H5/Python/File.cpp:
```diff
@@ -51,6 +52,19 @@ void bind_h5file() {
+         +[](const h5::H5File<h5::AccessType::ReadWrite>& f,
+             const std::string path) {
+           const auto& vol_file = f.get<h5::VolumeData>(path);
+             return &vol_file;
```
The indentation here looks a little funny :)

isaaclegred replied:
`clang-format` ted

nilsdeppe commented on src/IO/H5/Python/File.cpp:
```diff
@@ -51,6 +52,19 @@ void bind_h5file() {
+
+    .def("insert_vol",
+         +[](h5::H5File<h5::AccessType::ReadWrite>& f, const std::string path,
+             uint32_t version) {
```
`const uint32_t`

nilsdeppe commented on src/IO/H5/Python/VolumeData.cpp:
```diff
@@ -0,0 +1,71 @@
+
```
Don't need the blank line at the top of the file

nilsdeppe commented on src/IO/H5/Python/VolumeData.cpp:
```diff
@@ -0,0 +1,71 @@
+      // The method extension() is static
+      .staticmethod("extension")
+
+      .def("get_header", +[](h5::VolumeData& V) { return V.get_header(); })
```
`V` -> `volume_file` throughout. Clear names are worth an enormous amount in terms of having the code document itself :)

nilsdeppe commented on src/IO/H5/Python/VolumeData.cpp:
```diff
@@ -0,0 +1,71 @@
+
+      .def("get_header", +[](h5::VolumeData& V) { return V.get_header(); })
+
+      .def("get_version", +[](h5::VolumeData& V) { return V.get_version(); })
```
for these `get` functions can the `h5::VolumeData` be a `const&`? If not, it would be useful to add a comment right before `get_header` stating why.

nilsdeppe commented on src/IO/H5/Python/VolumeData.cpp:
```diff
@@ -0,0 +1,71 @@
+             return std_vector_to_py_list<size_t>(V.list_observation_ids());
+           })
+      .def("get_observation_value",
+           +[](h5::VolumeData& V, size_t observation_id) {
```
`const size_t` here and below

nilsdeppe commented on tests/Unit/IO/CMakeLists.txt:
```diff
@@ -33,3 +33,8 @@ spectre_add_python_test(
   "Unit.IO.H5.Python"
   "Test_H5.py"
   "unit;IO;H5;python")
+spectre_add_python_test(
```
can you add a blank line above the `spectre_add_python_test` so that the file is a bit easier to read at a quick glance?

nilsdeppe commented on src/IO/H5/Python/VolumeData.cpp:
```diff
@@ -0,0 +1,71 @@
+#include <boost/python.hpp>
+#include <boost/python/stl_iterator.hpp>
+#include <boost/python/suite/indexing/vector_indexing_suite.hpp>
+#include <hdf5.h>
```
what is `hdf5.h` being used for? I feel like this one should be something that is handled by the H5 wrappers.

isaaclegred replied:
Removed

nilsdeppe commented on src/IO/H5/Python/VolumeData.cpp:
```diff
@@ -0,0 +1,71 @@
+#include <vector>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Matrix.hpp"
```
What uses Matrix?

isaaclegred replied:
Also Removed

nilsdeppe commented on tests/Unit/IO/Test_VolumeData.py:
```diff
@@ -0,0 +1,93 @@
+        # these will be deleted and recreated for each test as needed.
+        self.file_name_w = "pythontest.h5"
+        # The other tests require external data, this file will not be changed
+        self.file_name_r = Informer.unit_test_path() + "IO/VolTestData.h5"
```
I think `sys.path` has some sort of function that safely concatenates paths. We should use that :)

isaaclegred replied:
@nilsdeppe I couldn't find anything in `sys.path` but it seems `os.path` has a function `join(base_dir, file_name)` which seems to be platform independent concatenation, is this what you meant?

nilsdeppe replied:
whoops, yes sorry :)

isaaclegred replied:
No Problem!

nilsdeppe commented on tests/Unit/IO/Test_VolumeData.py:
```diff
@@ -0,0 +1,93 @@
+        self.assertEqual(vol_file.get_header()[0:20], "#\n# File created on ")
+        h5_file.close()
+
+    # The other tests require external data, so they use the included h5 file
```
Could you provide the name of the h5 file (so it's more self-contained), and also how the file was generated?

isaaclegred replied:
I put a longer note at the top as well

nilsdeppe commented on tests/Unit/IO/Test_VolumeData.py:
```diff
@@ -0,0 +1,93 @@
+        expected_obs_ids = [836167400470363061, 1288279752782410135]
+        expected_obs_values = [0.03, 0.02]
+        self.assertItemsEqual(obs_ids, expected_obs_ids)
+        for i, obs_id in enumerate(expected_obs_ids):
```
Can we rely on ordering here or should we sort the lists?

isaaclegred replied:
Made it into a dictionary because it made more sense to me.

nilsdeppe commented on tests/Unit/IO/Test_VolumeData.py:
```diff
@@ -0,0 +1,93 @@
+                                 'InertialCoordinates_y',
+                                 'InertialCoordinates_z']
+        self.assertItemsEqual(tensor_comps, expected_tensor_comps)
+        tensor_data = vol_file.get_tensor_component(
```
Keep in mind I don't know exactly what's in the h5 file, but it would seem good to compare all components and to have not all the values be the same :)

isaaclegred replied:
Changed the test data and am now testing all of the components and they're all have different data.  One thing that is tough is that the DataVector wrapper doesn't implement any kind of "`almost_equal`" function, so I had to sort of clumsily convert everything to numpy arrays.

nilsdeppe replied:
That seems fine to me. Did you try using `np.asarray`? I'm not sure if it uses interators or what to convert, so it's a 50/50 chance that'll work. If it doesn't work could you add a comment to your description of `as_numpy_array` stating that `np.asarray` doesn't work?

nilsdeppe reviewed (changes requested):
I'm happy for you to squash. I haven't yet downloaded and looked at the h5 file, but the rest looks good :+1:

nilsdeppe commented on src/IO/H5/Python/File.cpp:
```diff
@@ -49,8 +50,22 @@ void bind_h5file() {
+
+      .def("get_vol",
+           +[](const h5::H5File<h5::AccessType::ReadWrite>& f,
+               const std::string path) {
```
Sorry I missed this last time, but the strings should be taken by `const std::string&` (reference at the end). Below as well in `insert_vol`

nilsdeppe reviewed (commented):
Looks good, just two tiny things to clarify the code a bit. I'm happy with this :)

isaaclegred commented:
@nilsdeppe It seems that `np.asarray` does work, so I've changed the code to reflect that.  I also fixed a typo in a test name and added a `const` in the `groups` function in `File.cpp` that probably should have been `const` the whole time.

nilsdeppe commented:
@isaaclegred great, thank you! This looks good. I'm happy for you to squash and rebase on develop :)

isaaclegred commented:
Hi @nilsdeppe  and @kidder, a couple of the travis tests were failing and I think it may have been because of the way I got the path in `tests/Unit/IO/Test_VolumeData.py` for `pythontest.h5`, so I changed that and rebased on develop, and am hoping the build will work this time.

nilsvu reviewed (approved):
Looks good 👍 I have only minor formatting comments, mostly whitespace, so won’t request those changes. If you end up making any more changes to this PR, please run autopep8 over the py file (generally a good idea).
